File: cli/go/cmd/prompd/uninstall.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
)

func handleUninstall() {
	if len(os.Args) < 3 {
		printUninstallUsage()
		os.Exit(1)
	}

	packages := []string{}
	globalUninstall := false

	// Parse arguments and flags
	for i := 2; i < len(os.Args); i++ {
		arg := os.Args[i]
		if arg == "-g" || arg == "--global" {
			globalUninstall = true
		} else if arg == "-h" || arg == "--help" {
			printUninstallUsage()
			return
		} else if !strings.HasPrefix(arg, "-") {
			packages = append(packages, arg)
		}
	}

	if len(packages) == 0 {
		fmt.Println("ERROR: No packages specified")
		printUninstallUsage()
		os.Exit(1)
	}

	// Uninstall each package
	for _, packageName := range packages {
		if err := uninstallPackage(packageName, globalUninstall); err != nil {
			fmt.Printf("ERROR: Failed to uninstall %s: %v\n", packageName, err)
			continue
		}

		location := "locally"
		if globalUninstall {
			location = "globally"
		}
		fmt.Printf("✓ Uninstalled %s %s\n", packageName, location)
	}
}

func uninstallPackage(packageName string, global bool) error {
	cacheDir := getCacheDir(global)

	// Find the package directory
	packageDir := filepath.Join(cacheDir, sanitizePackageName(packageName))

	// Check if package exists
	if _, err := os.Stat(packageDir); os.IsNotExist(err) {
		return fmt.Errorf("package not found in cache")
	}

	// Remove the package directory
	if err := os.RemoveAll(packageDir); err != nil {
		return fmt.Errorf("failed to remove package: %w", err)
	}

	return nil
}

func getCacheDir(global bool) string {
	homeDir, err := os.UserHomeDir()
	if err != nil {
		homeDir = "."
	}

	if global {
		return filepath.Join(homeDir, ".prompd", "cache", "global")
	}
	return filepath.Join(homeDir, ".prompd", "cache", "packages")
}

func sanitizePackageName(name string) string {
	// Convert @namespace/name to namespace-name for directory
	name = strings.ReplaceAll(name, "@", "")
	name = strings.ReplaceAll(name, "/", "-")
	return name
}

func printUninstallUsage() {
	fmt.Println(`Usage: prompd uninstall <packages...> [options]

Uninstall packages

Arguments:
  <packages...>         One or more package names to uninstall

Options:
  -g, --global          Uninstall packages globally

Examples:
  prompd uninstall @namespace/package
  prompd uninstall @namespace/package1 @namespace/package2
  prompd uninstall @namespace/package --global`)
}

File: typescript/src/commands/uninstall.ts
import { Command } from 'commander';
import chalk from 'chalk';
import * as fs from 'fs-extra';
import * as path from 'path';
import * as os from 'os';
import { PACKAGE_TYPE_DIRS, resolveToolDeployDir, TOOL_DEPLOY_DIRS } from '../types';

export function createUninstallCommand(): Command {
  const uninstallCommand = new Command('uninstall');

  uninstallCommand
    .description('Uninstall packages')
    .argument('<packages...>', 'Package names to uninstall')
    .option('-g, --global', 'Uninstall packages globally')
    .action(async (packages: string[], options) => {
      try {
        for (const packageName of packages) {
          const location = options.global ? 'globally' : 'locally';
          console.log(chalk.dim(`Uninstalling ${packageName} ${location}...`));

          const removed = await uninstallPackage(packageName, options.global);

          if (removed) {
            console.log(chalk.green(`Uninstalled ${packageName}`));
          } else {
            console.log(chalk.yellow(`Package ${packageName} not found`));
          }
        }
      } catch (error) {
        console.error(chalk.red(`Uninstall failed: ${error}`));
        process.exit(1);
      }
    });

  return uninstallCommand;
}

/**
 * Uninstall a package by searching across all type directories.
 * Also cleans up tool deployments if the package was a skill deployed to a tool.
 */
async function uninstallPackage(packageName: string, global: boolean): Promise<boolean> {
  const baseDir = global
    ? path.join(os.homedir(), '.prompd')
    : path.join(process.cwd(), '.prompd');

  let removed = false;

  // Search across all type directories (packages/, workflows/, skills/, templates/)
  for (const typeDir of Object.values(PACKAGE_TYPE_DIRS)) {
    const packageDir = path.join(baseDir, typeDir, packageName);

    if (await fs.pathExists(packageDir)) {
      // Before removing, check if this was a skill with tool deployments
      if (typeDir === 'skills') {
        await cleanupToolDeployments(packageName);
      }

      await fs.remove(packageDir);
      removed = true;
    }
  }

  // Also check legacy cache directory for backward compatibility
  const legacyCacheDir = path.join(baseDir, 'cache');
  const legacyPackageDir = path.join(legacyCacheDir, packageName);
  if (await fs.pathExists(legacyPackageDir)) {
    await fs.remove(legacyPackageDir);
    removed = true;
  }

  // Also check legacy global cache structure
  if (global) {
    const legacyGlobalDir = path.join(baseDir, 'cache', 'global', 'packages', packageName);
    if (await fs.pathExists(legacyGlobalDir)) {
      await fs.remove(legacyGlobalDir);
      removed = true;
    }
  }

  return removed;
}

/**
 * Clean up tool-native deployments for a skill package.
 * Checks each known tool's skill directory for a deployment marker (.prompd-source).
 */
async function cleanupToolDeployments(packageName: string): Promise<void> {
  for (const toolName of Object.keys(TOOL_DEPLOY_DIRS)) {
    const deployDir = resolveToolDeployDir(toolName);
    if (!deployDir) continue;

    const skillDeployDir = path.join(deployDir, packageName);
    const markerPath = path.join(skillDeployDir, '.prompd-source');

    // Only remove if the .prompd-source marker exists (confirms prompd deployed it)
    if (await fs.pathExists(markerPath)) {
      await fs.remove(skillDeployDir);
      console.log(chalk.dim(`  Cleaned up ${toolName} deployment: ${skillDeployDir}`));
    }
  }
}
